sortinghat/core/log.py

```diff
@@ -0,0 +1,152 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2014-2019 Bitergia
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+# Authors:
+#     Miguel Ángel Fernández <mafesan@bitergia.com>
+#
+
+
+import json
+import re
+import uuid
+
+import django.core.exceptions
+import django.db.utils
+
+from grimoirelab_toolkit.datetime import datetime_utcnow
+
+from .errors import AlreadyExistsError, ClosedTransactionError
+from .models import (Operation,
+                     Transaction)
+
+from .utils import validate_field
+
+
+class TransactionsLog:
+    """Class for logging transactions and operations related with the database.
+
+    Every object of this class is created using the `open` class method receiving
+    as a parameter the name of the method opening the transaction, which creates a
+    `Transaction` object in the database.
+
+    The method `log_operation` creates a new `Operation` objects linked to the
+    transaction which was generated when the object was instanced. This method
+    receives, among other parameters, the arguments from the method logging the
+    operation in a Python dict object which will be converted to a serialized
+    JSON.
+
+    The `close` method adds a `closed_at` timestamp (before calling this method,
+    this field has a `NULL` value in the DB) and it also sets to `True` the
+    `is_closed` flag.
+
+    :param trx: Transaction object generated with the class method `open`
+
+    :raises ClosedTransactionError: When trying to log an operation on a closed transaction
+    :raises TypeError: When the `op_type` is not an instance of `Operation.OpType` class
+    """
+    def __init__(self, trx):
+
+        self.trx = trx
+
+    @classmethod
+    def open(cls, name):
+        """Create a new transaction object and save it into the DB.
+
+        :param name: Name of the method opening the transaction
+
+        :returns: a new TransactionsLog object containing the generated Transaction object
+        """
+        validate_field('name', name)
+
+        tuid = uuid.uuid4().hex
+
+        trx = Transaction(tuid=tuid, name=name, created_at=datetime_utcnow())
```

Comment:
Instead of calling to `Transaction` call to `cls`. This will make possible to inherit from `Transaction`.

Reply:
My mistake, this is correctly done.